### PR TITLE
Set webroot-map from diverse flags, user interaction, and config files correctly

### LINKED ISSUE #1668: Several code paths (conf files, interactive domains, --csr) do not set webroot_map entries correctly
This may or may not be intended.

If you set `webroot-path` in your config and don't specify your domains, letsencrypt fails with the following error:
```
The webroot plugin is not working; there may be problems with your existing configuration.
The error was: PluginError('--webroot-path must be set',)
```
This is pretty confusing, both because `webroot-path` is set, and also because the code in question actually uses `webroot-map`, which is autogenerated in cli.py if both webroot and domains are specified upon invocation.

Previously this use pattern worked.

These changes were introduced in e1f0fcc and ffe6226e.

### PR DESCRIPTION
**Pull request by pde**
Closes #1668

### CODE REVIEW

**Comment by pde** [comment]
I've incorporated @filipochnik's PR #2241 into this one, since it touches many of the same webroot-map processing mechanisms.

File: letsencrypt/cli.py [hunk 1]
```diff
@@ -1282,19 +1294,36 @@ def __call__(self, parser, config, webroot, option_string=None):
...
+            # unless it was explicitly included in webroot_map
+            if webroot_path:
+                config.webroot_map.setdefault(domain, webroot_path[-1])
+
+
+class WebrootMapProcessor(argparse.Action): # pylint: disable=missing-docstring
+    def __call__(self, parser, config, webroot_map_arg, option_string=None):
+        webroot_map = json.loads(webroot_map_arg)
+        for domains, webroot_path in webroot_map.iteritems():
+            _process_domain(config, str(domains), [str(webroot_path)])
```
Comment (filipochnik) [inline letsencrypt/cli.py]: Right now something like `-d żąćł.com` results in a nice error "Internationalized domain names are not presently supported". `--webroot-map "{'żąćł.com': '/tmp'}"` on the other hand raises `UnicodeEncodeError`. Would be nice to have consistency between these two options.
Maybe moving `str(domains)` to `_process_domain` and wrapping it in a try clause?

`str(webroot_path)` raises an uncaught exception here as well, but utf-8 characters in webroot_path seem to be unsupported now anyway?

**Comment by filipochnik** [comment]
Other than this unicode thing LGTM :)

**Comment by pde** [comment]
Unicode issue fixed; merging.

**Comment by filipochnik** [comment]
@pde Does it work for you?
```
$ sudo letsencrypt certonly --staging --test-cert --text --webroot --email a@a.com --webroot-map '{"test.com": "/tmp"}'
An unexpected error occurred:
TypeError: initializer for ctype 'char *' must be a str or list or tuple, not unicode
Please see the logfiles in /var/log/letsencrypt for more details.
```
Traceback from logs:
```
2016-01-30 13:11:23,525:DEBUG:letsencrypt.cli:Exiting abnormally:
Traceback (most recent call last):
  File "/home/filip/SSD/letsencrypt/venv/bin/letsencrypt", line 9, in <module>
    load_entry_point('letsencrypt', 'console_scripts', 'letsencrypt')()
  File "/home/filip/SSD/letsencrypt/letsencrypt/cli.py", line 1484, in main
    return args.func(args, config, plugins)
  File "/home/filip/SSD/letsencrypt/letsencrypt/cli.py", line 644, in obtain_cert
    _auth_from_domains(le_client, config, domains)
  File "/home/filip/SSD/letsencrypt/letsencrypt/cli.py", line 422, in _auth_from_domains
    lineage = le_client.obtain_and_enroll_certificate(domains)
  File "/home/filip/SSD/letsencrypt/letsencrypt/client.py", line 283, in obtain_and_enroll_certificate
    certr, chain, key, _ = self.obtain_certificate(domains)
  File "/home/filip/SSD/letsencrypt/letsencrypt/client.py", line 264, in obtain_certificate
    csr = crypto_util.init_save_csr(key, domains, self.config.csr_dir)
  File "/home/filip/SSD/letsencrypt/letsencrypt/crypto_util.py", line 78, in init_save_csr
    csr_pem, csr_der = make_csr(privkey.pem, names)
  File "/home/filip/SSD/letsencrypt/letsencrypt/crypto_util.py", line 118, in make_csr
    value=", ".join("DNS:%s" % d for d in domains)
  File "/home/filip/SSD/letsencrypt/venv/lib/python2.7/site-packages/OpenSSL/crypto.py", line 651, in __init__
    extension = _lib.X509V3_EXT_nconf(_ffi.NULL, ctx, type_name, value)
TypeError: initializer for ctype 'char *' must be a str or list or tuple, not unicode
```
Seems like pyOpenSSL expects str.

**Comment by pde** [comment]
@filipochnik sorry just saw this note; yes, current git master works fine for me with a command like that.

Aside: `--staging` and `--test-cert` are synonyms, you don't need both!

**Comment by filipochnik** [comment]
Yeah, the issue was fixed by #2336 

Didn't know that about `--staging` and `--test-cert`, thanks.

**Pull request merged**